### PR TITLE
Redfish stream multipart uploads

### PR DESCRIPTION
**Pull request by joelrebel**
## What does this PR implement/change/remove?

The Redfish provider currently reads the firmware files into memory before a multipart upload,
this results in high memory consumption for the client and can be problematic for cases
where firmware installs have to be performed simultaneously.

This change was part of bmclib v1 https://github.com/bmc-toolbox/bmclib/pull/279 
but was somehow missed from being included in v2.

Tested with Dell R6515's, R7640.

heap profile - with streaming uploads
```
(pprof) top 20
Showing nodes accounting for 2059.99kB, 100% of 2059.99kB total
Showing top 20 nodes out of 29
      flat  flat%   sum%        cum   cum%
  520.04kB 25.24% 25.24%  1034.42kB 50.21%  encoding/json.(*Decoder).refill
  514.38kB 24.97% 50.21%   514.38kB 24.97%  compress/flate.NewReader
  513.56kB 24.93% 75.15%   513.56kB 24.93%  regexp/syntax.init
  512.01kB 24.85%   100%   512.01kB 24.85%  sync.(*Map).Swap
         0     0%   100%   514.38kB 24.97%  compress/gzip.(*Reader).Reset
         0     0%   100%   514.38kB 24.97%  compress/gzip.(*Reader).readHeader
         0     0%   100%   514.38kB 24.97%  compress/gzip.NewReader
         0     0%   100%  1546.43kB 75.07%  encoding/json.(*Decoder).Decode
         0     0%   100%  1034.42kB 50.21%  encoding/json.(*Decoder).readValue
         0     0%   100%   512.01kB 24.85%  encoding/json.(*decodeState).object
         0     0%   100%   512.01kB 24.85%  encoding/json.(*decodeState).unmarshal
         0     0%   100%   512.01kB 24.85%  encoding/json.(*decodeState).value
         0     0%   100%   512.01kB 24.85%  encoding/json.Unmarshal
         0     0%   100%   512.01kB 24.85%  encoding/json.cachedTypeFields
         0     0%   100%   512.01kB 24.85%  encoding/json.typeEncoder
         0     0%   100%   512.01kB 24.85%  encoding/json.typeFields
         0     0%   100%  1032.05kB 50.10%  github.com/stmcginnis/gofish/common.(*Entity).Get
         0     0%   100%  1032.05kB 50.10%  github.com/stmcginnis/gofish/common.CollectCollection.func1
         0     0%   100%   514.38kB 24.97%  github.com/stmcginnis/gofish/common.CollectList
         0     0%   100%   514.38kB 24.97%  github.com/stmcginnis/gofish/common.GetCollection
```

heap profile - without streaming uploads
```
(pprof) top 30
Showing nodes accounting for 240MB, 99.58% of 241MB total
Dropped 16 nodes (cum <= 1.21MB)
      flat  flat%   sum%        cum   cum%
     240MB 99.58% 99.58%      240MB 99.58%  bytes.growSlice
         0     0% 99.58%      240MB 99.58%  bytes.(*Buffer).Write
         0     0% 99.58%      240MB 99.58%  bytes.(*Buffer).grow
         0     0% 99.58%   240.50MB 99.79%  github.com/bmc-toolbox/bmclib/v2.(*Client).FirmwareInstall
         0     0% 99.58%   240.50MB 99.79%  github.com/bmc-toolbox/bmclib/v2/bmc.FirmwareInstallFromInterfaces
         0     0% 99.58%   240.50MB 99.79%  github.com/bmc-toolbox/bmclib/v2/bmc.firmwareInstall
         0     0% 99.58%   240.50MB 99.79%  github.com/bmc-toolbox/bmclib/v2/providers/redfish.(*Conn).FirmwareInstall
         0     0% 99.58%   240.50MB 99.79%  github.com/bmc-toolbox/bmclib/v2/providers/redfish.(*Conn).runRequestWithMultipartPayload
         0     0% 99.58%      240MB 99.58%  io.Copy (inline)
         0     0% 99.58%      240MB 99.58%  io.copyBuffer
         0     0% 99.58%   240.50MB 99.79%  main.main
         0     0% 99.58%      240MB 99.58%  mime/multipart.(*part).Write
         0     0% 99.58%      241MB   100%  runtime.main
```

### Checklist
- [X] Tests added
- [X] Similar commits squashed

## Description for changelog/release notes

```
- Redfish streaming multipart uploads
```

### CODE REVIEW

File: providers/redfish/firmware.go
Comment (mmlb) [inline providers/redfish/firmware.go:211]: small nit: wouldn't returning an error be better than panic'ing?

**Review by mattcburns — APPROVED** [review]

File: providers/redfish/firmware.go
Comment (mmlb) [inline providers/redfish/firmware.go:238]: Any clue what other kinds of readers might show up here? It might be nice to see if the implementation of reader might have a .Len method so we can avoid unnecessary reads.

```go
type Lenner interface {
	Len() int
}
...

if file, ok := reader.(*os.File); ok {
...
} else if lenner, ok := reader.(Lenner); ok {
...
} else {
// read the file to get size
}
```

File: providers/redfish/firmware.go
Comment (mmlb) [inline providers/redfish/firmware.go:215]: this func seems rather wasteful, you go through the effort of creating a form only to throw it away to then create another one later in the caller of this func with almost duplicate code... why not just return the form + size from this func?

Comment (joelrebel) [reply providers/redfish/firmware.go:215]: The `multipartPayloadSize()` function copies the contents of the multipart form into a `bytes.Buffer{}` to determine the actual form size. 
The`runRequestWithMultipartPayload()` uses a `io.PipeWriter` which means the `io.Copy()` blocks until whenever the http request is actually being sent - to pipe the update file contents. 

So I don't see a trivial way to determine the Content-Length without building the multipart form once before actually sending it, I'm open to any ideas thought.

File: providers/redfish/firmware.go
Comment (joelrebel) [reply providers/redfish/firmware.go:238]: Hmm, this might need some rethinking - since we also require the file name here,
```go
  if file, ok := reader.(*os.File); ok {
    // Add update file fields
    if _, err = form.CreateFormFile(key, filepath.Base(file.Name())); err != nil {
      return 0, body, err
    }
    ...
  }  
```

File: providers/redfish/firmware.go
Comment (joelrebel) [reply providers/redfish/firmware.go]: thanks, fixed in [6cf0168](https://github.com/bmc-toolbox/bmclib/pull/345/commits/6cf01686c513a1d94b75b69bed6e7689fd26e5ce)

File: providers/redfish/firmware.go
Comment (mmlb) [reply providers/redfish/firmware.go:218]: Aw yeah I see that now. Nope I can't really think of another option w/o checking interfaces for Len and/or Size and maybe avoid the read. :(

**Comment by joelrebel** [comment]
@mmlb whenever you're back, PTAL :)

**Review by ofaurax — DISMISSED** [review]
Waiting for @mmlb but I give my approval since it's needed for #346

File: providers/redfish/firmware.go
Comment (mmlb) [reply providers/redfish/firmware.go:239]: gotcha, is that described somewhere in this repo (docs maybe)? Either way it would be nice to have a check in this loop maybe to make it obvious to code readers.

Comment (joelrebel) [reply providers/redfish/firmware.go]: I took a slightly different route and simplified the implementation - [059b754](https://github.com/bmc-toolbox/bmclib/pull/345/commits/059b754ae098c55eb3b2466e5aaf849e131ea974)

Instead of expecting a payload `[]map[string]io.Reader` its now the `multipartPayload` struct type,
the loop is also unrolled and its simpler to follow the form payload being built.

File: providers/redfish/firmware.go
Comment (mmlb) [inline providers/redfish/firmware.go:138]: So I went looking up the calls stack and now I'm wondering, why does this need to be an io.Reader at all? Its always going to be a `[]byte` from `json.Marshal` in https://github.com/bmc-toolbox/bmclib/blob/a1b87e2ff47f5c05022a6cb8fddc6811e13d9a89/providers/redfish/firmware.go#L79-L87. I don't see why we want to make it a Reader and have to have all the complications of tee'ing and resetting. If its kept as a []byte then down below in multipartPayloadSize p
it'd be something like:

```go
if _, err = part.Write(payload.updateParameters); err != nil {
		return 0, body, err
}
```
right?

**Review by mmlb — COMMENTED** [review]
Agree that this looks much better! I think if we drop io.Reader for the updateParameters it will similarly make it simpler too.

File: providers/redfish/firmware.go
Comment (joelrebel) [reply providers/redfish/firmware.go]: thanks for feedback there, [08172f2](https://github.com/bmc-toolbox/bmclib/pull/345/commits/08172f2d91032140fb93567da8046fbcf5b76b0a)

**Review by mmlb — APPROVED** [review]
lgtm

**Pull request merged**
